pkg/catalog/manager/catalog_sync.go

```diff
@@ -64,51 +64,69 @@ func (m *Manager) Sync(key string, obj *v3.Catalog) (runtime.Object, error) {
 		return nil, err
 	}
 
-	// ensure the system catalog image cache exists
-	var systemCatalogImageCache *v1.ConfigMap
 	if catalog.Name == utils.SystemLibraryName {
-		var systemCatalogImageCacheCreated bool
-		systemCatalogImageCacheName := utils.GetCatalogImageCacheName(catalog.Name)
-		systemCatalogImageCache, err = m.ConfigMapLister.Get(namespace.System, systemCatalogImageCacheName)
-
-		// if the cache does not exist generate it
-		if err != nil && errors.IsNotFound(err) {
-			logrus.Debug("system catalog image cache configmap not found")
-
-			systemCatalogImageCache = &v1.ConfigMap{}
-			systemCatalogImageCache.Name = systemCatalogImageCacheName
-			systemCatalogImageCache.Namespace = namespace.System
-			err = image.CreateCatalogImageListConfigMap(systemCatalogImageCache, catalog)
-			if err != nil {
-				return nil, err
-			}
+		// ensure the system catalog image cache exists
+		forceUpdate := !isUpToDate(commit, catalog)
+		return nil, CreateOrUpdateSystemCatalogImageCache(catalog, m.ConfigMap, m.ConfigMapLister, false, forceUpdate)
+	}
 
-			systemCatalogImageCache, err = m.ConfigMap.Create(systemCatalogImageCache)
-			if err != nil && !errors.IsAlreadyExists(err) {
-				return nil, err
-			}
-			systemCatalogImageCacheCreated = true
-			logrus.Debug("system catalog image cache configmap created")
+	return nil, nil
+}
+
+func CreateOrUpdateSystemCatalogImageCache(systemCatalog *v3.Catalog, configMapInterface v1.ConfigMapInterface, configMapLister v1.ConfigMapLister, bundledMode bool, forceUpdate bool) (err error) {
```

Comment:
The original code utilized both a Lister (for Gets) and a Interface (for Create / Update) instead of just doing all calls via the Interface.

I debated making all calls go through the Interface instead to keep the function signature simple but chose to stick with this approach since I did not want to have to reach out to the k8s API in order to check if the ConfigMap exists; instead, looking through Rancher's cache of ConfigMaps via the Lister is sufficient and more efficient IMO.

Feel free to comment if you think I should drop the Lister in favor of just doing calls via the Interface!

Reply:
Using Lister for Gets and client for Create+Updates sounds good. If the logic is expected to always return object if it exists, we use Interface to Get if Lister returns NotFound but in this case can just use client to Create and no-op if AlreadyExists as you did. 👍🏼